README.md
# About the AWS Security Hub to Splunk integration.
The integration in this repository will send all findings in AWS Security Hub to Splunk for further analysis and correlation with relevant data sources (AWS CloudTrail, AWS CloudWatch, AWS Config, custom/on-prem data, etc.). Using automation provided by [Project Trumpet](https://github.com/splunk/splunk-aws-project-trumpet), AWS Security Hub Events are sent from AWS CloudWatch Events through a AWS Kinesis Data Firehose to a Splunk HTTP Event Collector. This is a much simpler path for "Getting Data In" than the older method of polling the AWS Security Hub API. 

# Setting up the AWS Security Hub to Splunk integration.

## Overview:
There are two parts to setting up the AWS Security Hub to Splunk integration.
	
#### Part 1: 
	
Follow the Trumpet instructions to build a Cloudformation template that will create a data pipeline for sending AWS Security Hub findings to Splunk 

#### Part 2: 
	
Load the `AWS_SecurityHub.spl` file in this repository containing the example dashboards for Security Hub. 

## Prerequisites:

- This step is **not** required. If AWS Security Hub is subscribed to AWS GuardDuty findings, check in the settings tab of the GuardDuty console that updated findings are sent the minimum of every 15 minutes through CloudWatch Events.

## Instructions:
#### Sending Security Hub Findings to Splunk:
        
Follow the [Trumpet Setup and Configuration guide](https://github.com/splunk/splunk-aws-project-trumpet)

#### Viewing Security Hub Findings in Splunk:

Install the `AWS_SecurityHub.spl` file in this repository containing the Splunk example app for AWS Security Hub. After Security Hub findings are received and indexed by Splunk, the dashboards will begin to populate.

## Troubleshooting

1.) Check if events are being sent to Splunk by searching `sourcetype="aws:securityhub"`. Make sure the time period is set to a range of time where you would expect events.

2.) Ensure the CloudFormation template was launched in an AWS region where AWS Security Hub is available.

3.) If there are no findings in the AWS Security Hub console, either wait for new findings, or generate sample findings from the GuardDuty -> Settings -> Generate Sample findings option.

4.) Refer to Project Trumpet's Troubleshooting Guide for more. 


CHANGELOG.md
## 0.2.1 - 10/1/2019
* migrate to prefered push GDI path over olld poller via Splunk Project: Trumoet


## 0.1.1 - 2/5/2019

* Hostname parser updated to support arbitrary HTTP Event Collector port configuration. Default port when none is specified in the provided SPLUNK_HEC_HOST is 443
* Error handling and logging improved
* SecurityHubToSplunkLambdaVerbose.zip file added to enable verbose logging when debugging
* Lambda timeout set to 840 seconds
* Lambda memory size set to 256

Archive/SecurityHubToSplunkLambdaVerbose/splunk_http_event_collector.py
"""splunk_http_event_collector.py
    Splunk HTTP event submission class

    Remember: Friends don't let friends send in non Common Information Model data: http://docs.splunk.com/Documentation/CIM/latest/User/Overview
        Please use CIM friendly field names when sending in data.
"""

__author__ = "george@georgestarcher.com (George Starcher)"

import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry

import json
import time
import socket
import threading
import uuid
import sys

is_py2 = sys.version[0] == '2'
if is_py2:
    import Queue as Queue
else:
    import queue as Queue

class http_event_collector:

    """
        Splunk HTTP Event Collector Class

        Keyword Arguments:
            token -- the Splunk HEC token value - required
            http_event_server -- the Splunk Server name or ip. Name must be network resolvable. - required
            input_type -- json or raw HEC type - provided at init (default json)
            host -- value to use as host field for events sent to Splunk (default the local system's hostname) 
            http_event_port -- Splunk HEC network port (default 8080)
            http_event_server_ssl -- boolean to set if Splunk HEC is using SSL (default True) 

        Attributes:
            debug -- debug boolean flag (default false) 
            SSL_verify -- boolean flag to force SSL certificate verification (default false)
            popNullFields -- boolean flag to pop null fields off payload prior to sending to Splunk (default false)
            index -- optional index name for HEC events (default None)
            sourcetype -- optional sourcetype name for HEC events (default None)
            server_uri -- computed property for HEC uri based on HEC type, raw metadata etc.

        Example Init:
            from splunk_http_event_collector import http_event_collector
            testeventJSON = http_event_collector("4D14F8D9-D788-4E6E-BF2D-D1A46441242E","localhost")

            For full usage example: https://github.com/georgestarcher/Splunk-Class-httpevent/blob/master/example.py
     """

    # Default batch max size to match splunk's default limits for max byte
    # See http_input stanza in limits.conf; note in testing I had to limit to 100,000 to avoid http event collector breaking connection
    # Auto flush will occur if next event payload will exceed limit
    maxByteLength = 100000
    threadCount = 10

    # An improved requests retry method from
    # https://www.peterbe.com/plog/best-practice-with-retries-with-requests
    # 503 added for endpoint busy
    # 408 added in case using HAproxy

    def requests_retry_session(self, retries=3,backoff_factor=0.3,status_forcelist=(408,500,502,503,504),session=None):
        session = session or requests.Session()
        retry = Retry(total=retries, read=retries, connect=retries, backoff_factor=backoff_factor, status_forcelist=status_forcelist, method_whitelist=frozenset(['HEAD', 'TRACE', 'GET', 'PUT', 'OPTIONS', 'DELETE', 'POST']))
        adapter = HTTPAdapter(max_retries=retry)
        session.mount('http://', adapter)
        session.mount('https://', adapter)
        return session

    def __init__(self,token,http_event_server,input_type='json',host="",http_event_port='8088',http_event_server_ssl=True):
        self.token = token
        self.debug = False
        self.SSL_verify = False
        self.http_event_server = http_event_server
        self.http_event_server_ssl = http_event_server_ssl
        self.http_event_port = http_event_port
        self.index = ""
        self.sourcetype = ""
        self.batchEvents = []
        self.currentByteLength = 0
        self.input_type = input_type
        self.popNullFields = False 
        self.flushQueue = Queue.Queue(0)
        for x in range(self.threadCount):
            t = threading.Thread(target=self._batchThread)
            t.daemon = True
            t.start()

        # TODO Need to fix for this to be meaningful when batching
        self.failureLock = threading.Lock()
        self.failedToSend = False
        
        if self.SSL_verify == False:
            requests.packages.urllib3.disable_warnings()
    
        # Set host to specified value or default to localhostname if no value provided
        if host:
            self.host = host
        else:
            self.host = socket.gethostname()

        if self.debug:
            print (self.token)
            print (self.server_uri)
            print (self.input_type)               

    @property
    def server_uri(self):

       # Build and set server_uri for http event collector
        # Defaults to SSL if flag not passed
        # Defaults to port 8088 if port not passed

        if self.http_event_server_ssl:
            protocol = 'https'
        else:
            protocol = 'http'

        if self.input_type == 'raw':
            input_url = '/raw?channel='+str(uuid.uuid1())
            if self.sourcetype: input_url = input_url+'&sourcetype='+self.sourcetype
            if self.index: input_url = input_url+'&index='+self.index
        else:
            input_url = '/event'
            if self.sourcetype or self.index: input_url = input_url+'?'
            if self.sourcetype: input_url = input_url+'sourcetype='+self.sourcetype+"&"
            if self.index: input_url = input_url+'index='+self.index+"&"

        server_uri = '%s://%s:%s/services/collector%s' % (protocol, self.http_event_server, self.http_event_port, input_url)
        return (server_uri)


    def sendEvent(self,payload,eventtime=""):
        """Method to immediately send an event to the http event collector"""

        if self.input_type == 'json':
            # If eventtime in epoch not passed as optional argument and not in payload, use current system time in epoch
            if not eventtime and 'time' not in payload:
                eventtime = str(round(time.time(),3))
                payload.update({'time':eventtime})

            # Fill in local hostname if not manually populated
            if 'host' not in payload:
                payload.update({"host":self.host})

        # send event to http event collector
        event = []
        if self.input_type == 'json':
            if self.popNullFields:
                payloadEvent = payload.get('event')
                payloadEvent = {k:payloadEvent.get(k) for k,v in payloadEvent.items() if v}
                payload.update({"event":payloadEvent})
            event.append(json.dumps(payload))
        else:
            event.append(str(payload))

        self.flushQueue.put(event)
        if self.debug:
            print ("Single Submit: Sticking the event on the queue.")
            print (event)
        self._waitUntilDone()

    def batchEvent(self,payload,eventtime=""):
        """
            Recommended Method to place the event on the batch queue. Queue will auto flush as needed.
        """

        if self.input_type == 'json':
            # Fill in local hostname if not manually populated
            if 'host' not in payload:
                payload.update({"host":self.host})

            # If eventtime in epoch not passed as optional argument and not in payload, use current system time in epoch
            if not eventtime and 'time' not in payload:
                eventtime = str(round(time.time(),3)) 
                payload.update({"time":eventtime})
            if self.popNullFields:
                payloadEvent = payload.get('event')
                payloadEvent = {k:payloadEvent.get(k) for k,v in payloadEvent.items() if v}
                payload.update({"event":payloadEvent})
            payloadString = json.dumps(payload)

        else:
            payloadString = str(payload)
            if not payloadString.endswith("\n"):
                payloadString=payloadString+"\n"


        payloadLength = len(payloadString)

        if ((self.currentByteLength+payloadLength) > self.maxByteLength or (self.maxByteLength - self.currentByteLength) < payloadLength):
            if self.debug:
                print ("Auto Flush: Sticking the batch on the queue.")
            self.flushQueue.put(self.batchEvents)
            self.batchEvents = []
            self.currentByteLength = 0

        self.batchEvents.append(payloadString)
        self.currentByteLength += payloadLength

    def _batchThread(self):
        """Internal Function: Threads to send batches of events."""
        
        while True:
            if self.debug:
                print ("Events received on thread. Sending to Splunk.")
            payload = " ".join(self.flushQueue.get())
            headers = {'Authorization':'Splunk '+self.token}
            # try to post payload twice then give up and move on
            try:
                r = self.requests_retry_session().post(self.server_uri, data=payload, headers=headers, verify=self.SSL_verify)
                if (r.json()['code'] != 0):
                    with self.failureLock:
                        self.failedToSend = True
            except Exception:
                with self.failureLock:
                    self.failedToSend = True
                pass

            if self.debug:
                try:
                    print (r.text)
                except:
                    pass
            self.flushQueue.task_done()
            
    def _waitUntilDone(self):
        """Internal Function: Block until all flushQueue is empty."""
        self.flushQueue.join()
        return


    def flushBatch(self):
        """Method called to force flushing of remaining batch events.
           Always call this method before exiting your code to send any partial batch queue.
        """

        if self.debug:
            print ("Manual Flush: Sticking the batch on the queue.")
        self.flushQueue.put(self.batchEvents)
        self.batchEvents = []
        self.currentByteLength = 0
        self._waitUntilDone()

SecurityHubToSplunkLambda/security_hub_to_splunk_lambda.py
import json
import datetime
import dateutil
import os
import sys
import time
from splunk_http_event_collector import http_event_collector

envLambdaTaskRoot = os.environ['LAMBDA_TASK_ROOT']
sys.path.insert(0, envLambdaTaskRoot + '/NewBotoVersion')
import boto3

# Gets the time 60 minutes ago
def freshEventTimestampGenerator():
	tm = datetime.datetime.utcfromtimestamp(time.time())
	return time.mktime((tm - datetime.timedelta(minutes=60)).timetuple())

# Gets the epoch time of a UTC timestamp in a Security Hub finding
def findingTimestampGenerator(finding_time):
	d = dateutil.parser.parse(finding_time)
	d.astimezone(dateutil.tz.tzutc())
	
	return time.mktime(d.timetuple())

# Gets 20 most recent findings from securityhub
def getFindings(session, securityhub, filters={}):
	return securityhub.get_findings(
	    Filters=filters
	)

# Gets 20 most recent findings from securityhub
def updateFindingNote(session, securityhub, existing_note, principal, filters={}):
	return securityhub.update_findings(
	    Filters=filters,
	    Note={
	    	'Text': 'SENT TO SPLUNK: %s' % existing_note,
	    	'UpdatedBy': principal
	    }
	)

# Gets 20 findings from securityhub using the NextToken from a previous request
def getFindingsWithToken(session, securityhub, token, filters={}):
	return securityhub.get_findings(
	    Filters=filters,
	    NextToken=token
	)

# Isolates host, port, and path from the provided host string. Default is port 443
def hostIsolator(host):
	parsed_port = 8088

	protocol_split = host.split('://')
	protocol = protocol_split[0]
	host_port_path = protocol_split[1]
	potential_host = host_port_path.split(':')

	# No port provided in url string
	if (len(potential_host) == 1):
		# Set default port
		if (protocol == 'https'):
			parsed_port = '443'
		potential_path = potential_host[0].split('/', 1)
		# No path provided in url string
		if (len(potential_path) == 1):
			parsed_host = potential_path[0]
			parsed_path = ''
		# Path provided in url string
		else:
			parsed_host = potential_path[0]
			parsed_path = '/' + potential_path[1]\
	# Port provided in url string
	else:
		parsed_host = potential_host[0]
		potential_port_path = potential_host[1].split('/', 1)
		# No path provided in url string
		if (len(potential_port_path) == 1):
			parsed_port = potential_port_path[0]
			parsed_path = ''
		# Path provided in url string
		else:
			parsed_port = potential_port_path[0]
			parsed_path = '/' + potential_port_path[1]

	return parsed_host, parsed_port, parsed_path

def lambda_handler(event, context):
	http_event_collector_token = os.environ['SPLUNK_HEC_TOKEN']
	http_event_collector_host, http_event_collector_port, http_event_collector_path = hostIsolator(os.environ['SPLUNK_HEC_HOST'])

	eventSenderJSON = http_event_collector(http_event_collector_token, http_event_collector_host, 'json', http_event_port=http_event_collector_port)
	# eventSenderJSON.debug = True
    
	session = boto3.session.Session()
	securityhub = session.client('securityhub',region_name=os.environ['AWS_REGION'])
	
	results = getFindings(session, securityhub)
	
	fresh_events_after_this_time = freshEventTimestampGenerator()
	fresh_events = True
	first_call = True
	sent_count = 0
	
	while ((first_call or 'NextToken' in results) and fresh_events):
		# Loop through all findings (20 by default) returned by Security Hub API call
		# If finding has the string "SENT TO SPLUNK" in the finding note, the event is not sent but
		# loop will continue.
		# Fresh events will be sent to Splunk over HTTP Event Collector (HEC), "SENT TO SPLUNK" will
		# be prefixed to the finding's note.
		# Break out of the loop when we have looked back across the last hour of events (based on the
		# finding's LastObservedAt timestamp)
		first_call = False

		for finding in results['Findings']:
			finding_timestamp = findingTimestampGenerator(finding['LastObservedAt'])
			already_sent = False
			existing_note = ''
			principal = 'SplunkSecurityHubLambda'
			
			if 'Note' in finding:
				if 'SENT TO SPLUNK:' in finding['Note']['Text']:
					already_sent = True
				else:
					existing_note = finding['Note']['Text']
					principal = finding['Note']['UpdatedBy']
			
			if (finding_timestamp > fresh_events_after_this_time and not already_sent):
				payload = {}
				payload.update({'sourcetype':'aws:securityhub'})
				payload.update({'event':json.dumps(finding)})
				
				filters = {
					'Id': [
				         { 
				            'Comparison': 'EQUALS',
				            'Value': finding['Id']
				         }
				      ],
				    'LastObservedAt': [
				         { 
				            'Start': finding['LastObservedAt'],
				            'End': finding['LastObservedAt']
				         }
				      ],
				}

				eventSenderJSON.sendEvent(payload)

				if not eventSenderJSON.failedToSend:
					print('Event successfully sent to Splunk')
					eventSenderJSON.failedToSend = False
					updateFindingNote(session, securityhub, existing_note, principal, filters)
					sent_count += 1
				else:
					print('Event NOT successfully sent to Splunk')
			else:
				fresh_events = False
				break
		if (fresh_events):
			results = getFindingsWithToken(session, securityhub, results['NextToken'])

	print '%s findings sent to Splunk: %s' % (sent_count, eventSenderJSON.server_uri)